compiler/rustc_codegen_llvm/src/back/write.rs

```diff
@@ -104,7 +106,10 @@ pub fn create_informational_target_machine(sess: &Session) -> &'static mut llvm:
         .unwrap_or_else(|err| llvm_err(sess.diagnostic(), err).raise())
 }
 
-pub fn create_target_machine(tcx: TyCtxt<'_>, mod_name: &str) -> &'static mut llvm::TargetMachine {
+pub fn create_target_machine(
+    tcx: TyCtxt<'_>,
+    mod_name: &str,
+) -> (&'static mut llvm::TargetMachine, *mut llvm::TargetMachineCppArgs) {
```

Comment:
So if you are returning these two pieces of data together, then you could pack them both into `TargetMachine` as well.

In my previous comment I assumed you would create `TargetMachineCppArgs` first, then use a reference to it when constructing `TargetMachine` to manage lifetimes.
But if destruction happens unsafely through raw pointers anyway, then you don't need two structures, `LLVMRustDisposeTargetMachine` can just dispose everything using correct order.

Reply:
I put the ptrs to llvm::TargetMachine and llvm::TargetMachineCppArgs together in a struct. Added a drop function so I user don't have to call the dispose functions & don't have to use unsafe code.

Reply:
I still don't understand why the second structure (`LLVMTargetMachineCppArgs`) is necessary.
`LLVMTargetMachine` already contains `Options.MCOptions.Argv0` and `Options.MCOptions.CommandLineArgs`, and can destroy them in `LLVMRustDisposeTargetMachine`, no need to keep them somewhere else.

Reply:
Ah I think I missunderstood your question.
The reason for extra struct is that MCTargetOptions stores a references to a string array instead of directly storing them. See : https://github.com/llvm/llvm-project/blob/1d9dcca9e3cd36cbaf0759a4b084c808d1aee044/llvm/include/llvm/MC/MCTargetOptions.h#L81C2-L82
The extra struct actually held the strings, so I could hand out references to it.

I suppose I can create an array on the heap and assign it to the ArrayRef. Then I delete it in the dispose function by getting the ptr from the ArrayRef (Probably exactly what you were suggesting). This feels a bit weird, but seems to be way less of a change.

I'll try that when I have time again.

Reply:
Got it working. Was able to entirely remove the helper struct.
@rustbot review